File: sistema_de_gestion_empresarial/poo_python_dinero/pooDinero/logic/Usuario.py
import re

class Usuario:
    def __init__(self, nombre, edad, dni=None):
        self.nombre = nombre
        self.edad = edad
        self.dni = None
        if dni:
            if not self.set_dni(dni):
                print("DNI incorrecto")

    def set_dni(self, dni):#validacion dni
        if not isinstance(dni, str):
            return False

        dni = dni.upper().strip()#.strip() --> quitar espacios
        patron = r"^\d{8}-?[A-Z]$"

        if not re.match(patron, dni):
            return False

        self.dni = dni.replace("-", "") #eliminar el guion para almacenarlo
        return True

    def __str__(self):
        return (f"Nombre: {self.nombre}\n"
                f"Edad: {self.edad}\n"
                f"DNI: {self.dni}")

File: desarrollo_de_interfaces/dinero_extra/DineroExtra/Usuario.cs
using System.Text.RegularExpressions;

namespace TiendaGastos
{
    /// <summary>
    /// Clase que representa al usuario de la aplicación, con nombre, edad y DNI validado.
    /// </summary>
    public class Usuario
    {
        public string Nombre { get; set; }
        public int Edad { get; set; }
        public string Dni { get; private set; }

        public Usuario(string nombre, int edad)
        {
            Nombre = nombre;
            Edad = edad;
            Dni = null;
        }

        /// <summary>
        /// Establece el DNI si cumple con el formato correcto (8 dígitos y una letra final).
        /// Permite opcionalmente un guion entre los números y la letra.
        /// </summary>
        public bool SetDni(string dni)
        {
            string patron = @"^\d{8}-?[A-Z]$";
            if (Regex.IsMatch(dni, patron))
            {
                // Eliminamos el guion si existe
                Dni = dni.Replace("-", "");
                return true;
            }
            else
            {
                return false;
            }
        }

        public override string ToString()
        {
            string dniStr = Dni != null ? Dni : "No asignado";
            return $"Usuario: {Nombre}, Edad: {Edad}, DNI: {dniStr}";
        }


    }
}

